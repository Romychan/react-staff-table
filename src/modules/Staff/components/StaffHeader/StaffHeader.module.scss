.header {
  @apply flex justify-between items-center mb-4;
}

.input {
  @apply max-w-xs;
}

.dropdowns {
  @apply flex;
}

.dropdown {
  @apply w-48 mr-2 last:mr-0;
}
