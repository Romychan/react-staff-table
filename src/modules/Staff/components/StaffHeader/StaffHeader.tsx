import { Dispatch, SetStateAction } from 'react';

import { COMPANY_FILTER, ROLE_FILTER } from '../../constants/filters';

import { IQueryParams } from '@modules/Staff/interfaces';

import { useDebounce } from '@hooks/useDebounce';

import { Dropdown } from '@ui/Dropdown/Dropdown';
import { DropdownItem } from '@ui/DropdownItem/DropdownItem';
import { Input } from '@ui/Input/Input';

import styles from './StaffHeader.module.scss';

interface IStaffHeader {
  setQueryParams: Dispatch<SetStateAction<IQueryParams>>;
}

export const StaffHeader = ({ setQueryParams }: IStaffHeader) => {
  const handleSearchDebounce = useDebounce<string>((newValue) =>
    setQueryParams((prev) => ({ ...prev, _page: '1', q: newValue })),
  );

  return (
    <div className={styles.header}>
      <Input
        className={styles.input}
        placeholder="Search..."
        onChange={(event) => handleSearchDebounce(event.target.value)}
      />

      <div className={styles.dropdowns}>
        <Dropdown
          onChange={(value) =>
            setQueryParams((prev) => ({
              ...prev,
              _page: '1',
              company_like: value,
            }))
          }
          className={styles.dropdown}
        >
          {COMPANY_FILTER.map((item) => (
            <DropdownItem key={item.value} value={item.value}>
              {item.label}
            </DropdownItem>
          ))}
        </Dropdown>

        <Dropdown
          onChange={(value) =>
            setQueryParams((prev) => ({
              ...prev,
              _page: '1',
              role_like: value,
            }))
          }
          className={styles.dropdown}
        >
          {ROLE_FILTER.map((item) => (
            <DropdownItem key={item.value} value={item.value}>
              {item.label}
            </DropdownItem>
          ))}
        </Dropdown>
      </div>
    </div>
  );
};
