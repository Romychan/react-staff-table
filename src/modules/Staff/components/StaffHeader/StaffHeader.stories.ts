import type { Meta, StoryObj } from '@storybook/react';

import { StaffHeader } from './StaffHeader';

/** This component is used together with `Staff` to display data filtering. */
const meta: Meta<typeof StaffHeader> = {
  title: 'Modules/StaffHeader',
  component: StaffHeader,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: { setQueryParams: { action: 'setQueryParams' } },
};

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
