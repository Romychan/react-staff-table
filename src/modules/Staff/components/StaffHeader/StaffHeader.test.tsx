import {
  fireEvent,
  screen,
  render,
  RenderResult,
} from '@testing-library/react';
import { vi } from 'vitest';

import { StaffHeader } from './StaffHeader';

describe('StaffHeader', () => {
  let container: RenderResult;
  const handleChange = vi.fn();

  beforeEach(() => {
    container = render(<StaffHeader setQueryParams={handleChange} />);
  });

  it('should render correctly', () => {
    const companySelectElement = screen.getByText(/All Company/i);
    const roleSelectElement = screen.getByText(/All Role/i);
    const inputElement = screen.getByTestId('input-field');

    expect(companySelectElement).toBeInTheDocument();
    expect(roleSelectElement).toBeInTheDocument();
    expect(inputElement).toBeInTheDocument();
  });

  it('should working correctly select company', () => {
    const companySelectElement = screen.getByText(/All Company/i);

    fireEvent.click(companySelectElement);

    const companySelectItem = screen.getByText(/Google/i);

    fireEvent.click(companySelectItem);

    expect(companySelectItem).not.toBeInTheDocument();

    expect(screen.getByText(/Google/i)).toBeInTheDocument();
  });

  it('should working correctly select role', () => {
    const roleSelectElement = screen.getByText(/All Role/i);

    fireEvent.click(roleSelectElement);

    const roleSelectItem = screen.getByText(/Manager/i);

    fireEvent.click(roleSelectItem);

    expect(roleSelectItem).not.toBeInTheDocument();

    expect(screen.getByText(/Manager/i)).toBeInTheDocument();
  });

  it('should working correctly search input', () => {
    const inputElement = screen.getByTestId<HTMLInputElement>('input-field');
    handleChange.mockImplementation((e) => {
      expect(e.target.value).toBe('test');
    });

    fireEvent.change(inputElement, { target: { value: 'test' } });

    expect(inputElement.value).toBe('test');
    expect(handleChange).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
