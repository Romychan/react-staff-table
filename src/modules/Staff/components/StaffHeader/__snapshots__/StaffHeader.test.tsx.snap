// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StaffHeader > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_header_2da5b5"
      >
        <input
          class="_input_76cf42 _input_2da5b5"
          data-testid="input-field"
          placeholder="Search..."
          value=""
        />
        <div
          class="_dropdowns_2da5b5"
        >
          <div
            class="_container_b2c890 _dropdown_2da5b5"
          >
            <div
              class="_dropdown_b2c890"
              data-testid="dropdown-container"
            >
              <button
                class="_title_b2c890"
                data-testid="dropdown-header"
              >
                All Company
                <svg
                  class="icon icon-chevron _icon_b2c890"
                  data-testid="icon"
                  height="8"
                  stroke=""
                  width="8"
                >
                  <use
                    xlink:href="/src/assets/img/general/icons.svg#chevron"
                  />
                </svg>
              </button>
            </div>
          </div>
          <div
            class="_container_b2c890 _dropdown_2da5b5"
          >
            <div
              class="_dropdown_b2c890"
              data-testid="dropdown-container"
            >
              <button
                class="_title_b2c890"
                data-testid="dropdown-header"
              >
                All Role
                <svg
                  class="icon icon-chevron _icon_b2c890"
                  data-testid="icon"
                  height="8"
                  stroke=""
                  width="8"
                >
                  <use
                    xlink:href="/src/assets/img/general/icons.svg#chevron"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_header_2da5b5"
    >
      <input
        class="_input_76cf42 _input_2da5b5"
        data-testid="input-field"
        placeholder="Search..."
        value=""
      />
      <div
        class="_dropdowns_2da5b5"
      >
        <div
          class="_container_b2c890 _dropdown_2da5b5"
        >
          <div
            class="_dropdown_b2c890"
            data-testid="dropdown-container"
          >
            <button
              class="_title_b2c890"
              data-testid="dropdown-header"
            >
              All Company
              <svg
                class="icon icon-chevron _icon_b2c890"
                data-testid="icon"
                height="8"
                stroke=""
                width="8"
              >
                <use
                  xlink:href="/src/assets/img/general/icons.svg#chevron"
                />
              </svg>
            </button>
          </div>
        </div>
        <div
          class="_container_b2c890 _dropdown_2da5b5"
        >
          <div
            class="_dropdown_b2c890"
            data-testid="dropdown-container"
          >
            <button
              class="_title_b2c890"
              data-testid="dropdown-header"
            >
              All Role
              <svg
                class="icon icon-chevron _icon_b2c890"
                data-testid="icon"
                height="8"
                stroke=""
                width="8"
              >
                <use
                  xlink:href="/src/assets/img/general/icons.svg#chevron"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
